### PR TITLE
sensors: Add driver for Vischay VCNL36825T Proximity Sensor

### PR DESCRIPTION
**Pull request by deveritec-jusc**
Base driver without power management and interrupt support.

Tested on a custom board based on Nordic nRF 52840 with the sensor used for proximity measurements against an immobile target.

### CODE REVIEW

File: dts/bindings/sensor/vishay,vcnl36825t.yaml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      - "auto"
+      - "force"
+    description: |
+      Mode of operation.
+      In "auto" mode the sensor performs sampling the continuously,
+      in "force" mode the sampling is performed on every fetch command.
+
+  measurement-period:
+    type: int
+    default: 40
```
Comment (MaureenHelm) [inline dts/bindings/sensor/vishay,vcnl36825t.yaml:26]: The default needs to be justified in the description. Is it a register reset value?

This needs to be addressed in the other properties as well.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Needs to be rebased

File: dts/bindings/sensor/vishay,vcnl36825t.yaml (hunk 1 above)
Comment (deveritec-rosc) [reply dts/bindings/sensor/vishay,vcnl36825t.yaml:26]: added the justification. Is this okay?

**Comment by deveritec-rosc** [comment]
@MaureenHelm: I finished trigger support today, should I add the code to this pull request, or is a separate more suitable?

File: dts/bindings/sensor/vishay,vcnl36825t.yaml (hunk 1 above)
Comment (MaureenHelm) [reply dts/bindings/sensor/vishay,vcnl36825t.yaml:28]: Yes, thanks

**Review by MaureenHelm — DISMISSED** [review]
Needs to be rebased

**Comment by MaureenHelm** [comment]
> I finished trigger support today, should I add the code to this pull request, or is a separate more suitable?

Separate, please

File: drivers/sensor/vcnl36825t/vcnl36825t.c [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+				       VCNL36825T_PS_ST_STOP);
+		if (rc < 0) {
+			return rc;
+		}
+
+		rc = vcnl36825t_update(&config->i2c, VCNL36825T_REG_PS_CONF1, VCNL36825T_PS_ON_MSK,
+				       VCNL36825T_PS_OFF);
+		if (rc < 0) {
+			return rc;
+		}
```
Comment (avisconti) [inline drivers/sensor/vcnl36825t/vcnl36825t.c:95]: After a suspend/resume are all reg values preserved, correct? I'm assuming it is correct as the proximity sensor should not be powered-off.

**Review by avisconti — CHANGES_REQUESTED** [review]

File: drivers/sensor/vcnl36825t/vcnl36825t.c (hunk 2 above)
Comment (deveritec-rosc) [reply drivers/sensor/vcnl36825t/vcnl36825t.c:95]: Registers should be preserved according to the datasheet.

Setting `VCNL36825T_PS_OFF` stops the bias current (according to the datasheet), but the sensor stays powered without performing any operation.

File: drivers/sensor/vcnl36825t/vcnl36825t.h [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+/* PS_CONF1 */
+#define VCNL36825T_PS_ON_POS  1
+#define VCNL36825T_PS_CAL_POS 7
+
+#define VCNL36825T_PS_ON_MSK GENMASK(1, 1)
+
+#define VCNL36825T_PS_ON  (1 << VCNL36825T_PS_ON_POS)
+#define VCNL36825T_PS_OFF (0 << VCNL36825T_PS_ON_POS)
+
+#define VCNL36825T_PS_CAL (1 << VCNL36825T_PS_CAL_POS)
```
[the remark this answers is not in the document]
Comment (deveritec-rosc) [reply drivers/sensor/vcnl36825t/vcnl36825t.h:46]: I just used clang-format here. Actually, I am fine with both. I skim through the other drivers and would adapt to them. :)

File: drivers/sensor/vcnl36825t/Kconfig [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# VCNL36825T Proximity Sensor configuration options
+
+# Copyright (c) 2023 Juliane Schulze, deveritec Gmbh
```
[the remark this answers is not in the document]
Comment (deveritec-rosc) [reply drivers/sensor/vcnl36825t/Kconfig]: done

File: drivers/sensor/vcnl36825t/vcnl36825t.c [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+
+	(void)pm_device_state_get(dev, &state);
+	if (state != PM_DEVICE_STATE_ACTIVE) {
+		return -EBUSY;
+	}
+#endif
+
+	switch (chan) {
+	case SENSOR_CHAN_ALL:
+	case SENSOR_CHAN_PROX:
```
[the remark this answers is not in the document]
Comment (deveritec-rosc) [reply drivers/sensor/vcnl36825t/vcnl36825t.c:124]: done

File: dts/bindings/sensor/vishay,vcnl36825t.yaml [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    enum:
+      - 10
+      - 20
+      - 40
+      - 80
+      - 160
+      - 320
```
[the remark this answers is not in the document]
Comment (deveritec-rosc) [reply dts/bindings/sensor/vishay,vcnl36825t.yaml]: done

File: dts/bindings/sensor/vishay,vcnl36825t.yaml [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    default: "1"
+    enum:
+      - "1"
+      - "1.5"
+      - "2"
+      - "2.5"
+      - "3"
+      - "3.5"
+      - "4"
+      - "8"
```
[the remark this answers is not in the document]
Comment (deveritec-rosc) [reply dts/bindings/sensor/vishay,vcnl36825t.yaml]: done

**Review by avisconti — APPROVED** [review]
LGTM

Can you just add in this PR description how you tested it? (board, sample, etc).

**Comment by deveritec-rosc** [comment]
I tested the driver on a custom board based on a Nordic nRF 52840 with this sensor used for proximity measurements which we need to detect proximity against an immobile target.

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**